File: CPP/Data_Structures/BinarySearchTree.cpp
#include <iostream>
using namespace std;

struct node
{
    int value;
    node *left, *right;
};

node *createNode(int value)
{
    node *newNode = new node;
    newNode->value = value;
    newNode->left = newNode->right = NULL;
    return newNode;
}

node *insertBST(node *root, int value)
{
    if (root == NULL)
    {
        return createNode(value);
    }
    if (value < root->value)
        root->left = insertBST(root->left, value);
    else
        root->right = insertBST(root->right, value);
}

void printPreoder(node *root)
{
    if (root == NULL)
        return;
    cout << root->value << ' ';
    printPreoder(root->left);
    printPreoder(root->right);
}

void printInorder(node *root)
{
    if (root == NULL)
        return;
    printInorder(root->left);
    cout << root->value << ' ';
    printInorder(root->right);
}

void printPostorder(node *root)
{
    if (root == NULL)
        return;
    printPostorder(root->left);
    printPostorder(root->right);
    cout << root->value << ' ';
}

int main()
{
    node *root;
    root = insertBST(root, 4);
    insertBST(root, 2);
    insertBST(root, 1);
    insertBST(root, 3);
    insertBST(root, 4);
    insertBST(root, 5);
    insertBST(root, 6);
    printInorder(root);
    return 0;
}

File: Python/Data_Structures/BinarySearchTree.py
class Node(object):
    def __init__(self, value):
        self.value = value
        self.left = self.right = None


def insertBST(root, value):
    if root == None:
        root = Node(value)

    else:
        if value < root.value:
            if root.right is None:
                root.right = Node(value)
            else:
                insertBST(root.right, value)

        else:
            if root.left is None:
                root.left = Node(value)
            else:
                insertBST(root.left, value)


def preorderTraversal(root):

    if root == None:
        return
    print(root.value, end=' ')
    preorderTraversal(root.left)
    preorderTraversal(root.right)


def inorderTraversal(root):

    if root == None:
        return
    inorderTraversal(root.left)
    print(root.value, end=' ')
    inorderTraversal(root.right)


def postorderTraversal(root):

    if root == None:
        return
    postorderTraversal(root.left)
    postorderTraversal(root.right)
    print(root.value, end=' ')

# Test Code


root = Node(4)
insertBST(root, 2)
insertBST(root, 1)
insertBST(root, 3)
insertBST(root, 4)
insertBST(root, 5)
insertBST(root, 6)
print('Preorder Traversal : ', end='')
preorderTraversal(root)
print()
print('Inorder Traversal : ', end='')
inorderTraversal(root)
print()
print('Postorder Traversal : ', end='')
postorderTraversal(root)
print()
